at4793-rk3994/project.md

```diff
@@ -0,0 +1,63 @@
+  - Ramneek Kaur <rk3994@nyu.edu>
+---
+
+> In what follows instructions (including this one) are in blockquotes, while
+> text not in blockquotes is a sample. You should of course replace the title
+> author metadata that shows up above with appropriate values.
+>
+> You should start your proposal by stating a problem statement or an overall
+> goal  as below.
+
+For our Big Data and Machine Learning final project we plan to present a deep dive analysis and comparison between top monitoring and reporting open source solutions available for Kubernetes clusters.
+
+# Motivation
+> Next you should talk about why, citing sources for your beliefs.
+
+The DevOps conceptual practice forces us to break significant problems into smaller problems. Microservices fit perfectly here, where small services build-up a component, and these components make up an application. The microservice architecture utilizes small teams to develop functional components one by one. By using a microservice architecture, it’s easy to make it faster, scale up and down without impacting the whole system. It’s also better for fault tolerance and is platform and language agnostic. Everything has pros and cons and for microservice architecture, it’s hard to maintain testing and monitoring.
+
+In microservice architecture, we deploy microservices as a container and for orchestration of containers, we rely heavily on tools like Kubernetes or Docker Swarm. Once deployed, microservice architecture has thousands of services talking to each other through networking, which can make it very challenging to monitor. We have to monitor independent services more than service to service communication. Kubernetes is widely growing in the market and it has a large rapidly growing ecosystem. Over the time many organizations are investing their resources in Kubernetes. Since it is open source, there are numerous plugins and tools available for analysis and reporting of the clusters. There are no definite comparisons between all of these tools and plugins available and neither do these organisations have time and resources to explore each of them individually. 
+
+Resources Used:
+https://logz.io/blog/kubernetes-monitoring/
+
+# Proposed Solution
+> Next you should talk about how you plan to address the problem. 
+> This should include information about what infrastructure (e.g., CloudLab
+> resources) you plan to use, how you plan to make progress, and how you plan to
+> evaluate your view.
+
+* Firstly, we will set up a working 3 node kubernetes cluster.
+* Second, we plan to run varying workloads on the cluster.
+* Third, we will implement the few widely known tools individually.
+* At last, we aim to analyse and compare all the tools based on the following parameters:
+  * Ease of implementation.
+  * User Experience.
+  * Level of monitoring available: cluster level, node level and pod level.
+  * The time at which the user is notified before a potential failure.
+  * Maintenance Schedules
```

Comment:
Why do the tools have an impact on this?

Reply:
Tools generally don't have an impact on this. Different tools might provide different levels of monitoring. For example, there might be a tool that provides more intricate level of monitoring as compared to other tools.

Reply:
I am not sure this answers my question.

Reply:
So what we are trying to explain here is that , monitoring tools can have limitations on till what level on the cluster they can provide CPU usage/ memory usage etc. info.
For example:

1. cAdvisor tool is kind of limited in terms of providing these stats only on the container level
2. on the other hand, a more powerful tool like grafana , can provide these information for each pod also.

Reply:
My precise question is on the use of the term "Maintenance schedules". Why does monitoring granularity, which is what your answer focuses on, have an impact on maintenance?

Reply:
By Maintainence Schedules what we mean is that , since all these tools are open source they are maintained by a community and hence would keep having newer versions available. Since organizations need proper planning before hand for updating these tools with newer versions we thought that it would be useful to provide an estimate analysis on how frequently newer versions of the product are getting released and how frequently the older versions are becoming unsupported.